File: 922.sort-array-by-parity-ii/solution.kt
class Solution {
    fun sortArrayByParityII(nums: IntArray): IntArray {
        val (even, odd) = nums.partition{ it % 2 == 0 }
        return IntArray(nums.size) { if (it % 2 == 0) even[it / 2] else odd[it / 2] }
    }
}

File: 922.sort-array-by-parity-ii/solution.scala
object Solution {
    def sortArrayByParityII(A: Array[Int]): Array[Int] = {
        val p = A.partition(_ % 2 == 0)
        val s = for {
            (even, odd) <- p._1 zip p._2
            r <- List(even, odd)
        } yield r
        s.toArray
    }
}